case-lib/lib.sh

```diff
@@ -253,6 +253,50 @@ func_lib_get_tplg_path()
     return 0
 }
 
+# We must not quote SOF_ALSA_OPTS and disable SC2086 below for two reasons:
+#
+# 1. We want to support multiple parameters in a single variable, in
+#    other words we want to support this:
+#
+#      SOF_ALSA_OPTS="--foo --bar"
+#      aplay $SOF_ALSA_OPTS ...
+#
+# 2. aplay does not ignore empty arguments anyway, in other words this
+#    does not work anyway:
+#
+#      SOF_ALSA_OPTS=""
+#      aplay "$SOF_ALSA_OPTS" ...
+#
+# This is technically incorrect because it means our SOF_ALSA_OPTS
+# cannot support whitespace, for instance this would be split in two
+# options: --bar="aaaa and bbbb"
+#
+#      SOF_ALSA_OPTS='--foo --bar="aaaa bbbb"'
+#
+# To do this "correctly" SOF_ALSA_OPTS should be an array.
+# From https://mywiki.wooledge.org/BashGuide/Arrays "The only safe way
+# to represent multiple string elements in Bash is through the use of
+# arrays."
+# However 1. it would complicate the user interface 2. We don't seem to
+# need arguments with whitespace or globbing characters.
+
+# Default value, can be overriden by the environment. Empty string does
+# override.
+: "${SOF_ALSA_OPTS=--fatal-errors}"
+
+aplay_opts()
+{
+    dlogc "aplay $SOF_ALSA_OPTS $SOF_APLAY_OPTS $*"
+    # shellcheck disable=SC2086
+    aplay $SOF_ALSA_OPTS $SOF_APLAY_OPTS "$@"
```

Comment:
we also want to test the return status since it contains an error code when it fails. If you just stop I think the rest of the script will probably think nothing bad happened.

Reply:
The vast majority of the time `set -e` (errexit #312) will immediately abort the script, demonstration:

```shell
$ SOF_ALSA_OPTS=--bogus ./test-case/check-playback.sh

2020-11-06 19:05:09 UTC [COMMAND] sof-tplgreader.py /lib/firmware/intel/sof-tplg/sof-apl-pcm512x.tplg -f 'type:playback & id:any' -b ' pcm:HDA Digital' -s 0 -e
2020-11-06 19:05:09 UTC [INFO] ===== Testing: (Round: 1/1) (PCM: Port5 [hw:0,0]<playback>) (Loop: 1/3) =====
2020-11-06 19:05:09 UTC [COMMAND] aplay --bogus  -Dhw:0,0 -r 48000 -c 2 -f S16_LE -d 10 /dev/zero -v -q
aplay: unrecognized option '--bogus'
Try `aplay --help' for more information.
2020-11-06 19:05:09 UTC [ERROR] aplay on PCM hw:0,0 failed at 1/3.
2020-11-06 19:05:09 UTC [ERROR] Starting func_exit_handler(), exit status=1, FUNCNAME stack:
2020-11-06 19:05:09 UTC [ERROR]  die()  @  ./test-case/../case-lib/lib.sh
2020-11-06 19:05:09 UTC [ERROR]  main()  @  ./test-case/check-playback.sh:99
```
Same with `./test-case/check-capture.sh`

The other test `SOF_ALSA_OPTS=--bogus ./test-case/multiple-pipeline-capture.sh` doesn't stop immediately because `multiple-pipeline-capture.sh` backgrounds the `aplay` and `arecord` commands. Testing the error code inside this function would add complexity and not make any difference because the entire function is backgrounded.

What makes a difference with backgrounded commands is to collect ASAP their exit status with `wait $aplayPID`. This is something sof-test rarely does for now but that I systematically requested in every code review involving backgrounded command(s). `wait $aplay_optsPID` will collect the exit status of the last command in `aplay_opts()` function which is of course `aplay`. `set -e` applies to `wait` exactly the same and if desired it can be overriden with a better message (#312) the same.

Even without considering backgrounding issues, the caller of `aplay_opts` caller can generally provide more useful and relevant context if a custom error message is desired.